components/tidb_query_executors/src/runner.rs

```diff
@@ -438,14 +445,24 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
 
             let mut chunk = Chunk::default();
 
-            let (drained, record_len) = self.internal_handle_request(
+            let (drained, record_len, quota_delay) = self.internal_handle_request(
                 false,
                 batch_size,
                 &mut chunk,
                 &mut warnings,
                 &mut ctx,
             )?;
 
+            if !quota_delay.is_zero() {
```

Comment:
IMO you should put these boilerplate codes inside the limiter. E.g. Returns a new `Throttle` object instead of raw duration. Maybe pass in a reference to the prometheus `IntCounter` to get around the metrics stuff.

Reply:
fixed your first part

Reply:
I was thinking something like this, the timing stuff is hidden:
```rust
let throttle = limiter.new_throttle();
/* do the work */
throttle.read_bytes(stats.read);
throttle.read_bytes(stats.write);
throttle.async_consume(&TXN_COMMAND_THROTTLE_TIME_COUNTER_VEC_STATIC.get(CMD)).await;
```

Reply:
most agree but just async_consume.await will bring unnecessary switch. How about like this: 
```
let sample = limiter.new_sample();
/* do the work */
sample.read_bytes(stats.read);
sample.write_bytes(stats.write);
if sample.is_throttled() {
  limiter.async_consume(sample, &TXN_COMMAND_THROTTLE_TIME_COUNTER_VEC_STATIC.get(CMD)).await;
}
```

Reply:
You can do the conditional wait inside `async_consume`.

Reply:
after bench more voluntary switch if use this style